### PR TITLE
Create zip files based on metadata

### LINKED ISSUE #254: NPM script that packs ZIP files for publication
- [x] copy files from `docs` into folder with stage and publication date
- [x] rename md/html/pdf to follow OASIS naming conventions -> use `meta.filename`
- [x] copy additional artifacts from other repositories?
- [x] pack zip file without operating-system-specific stuff

Note: the publication date is not needed in the zip file name, the stage is sufficient

### PR DESCRIPTION
**Pull request by HeikoTheissen**
New metadata properties:
- `archivename` and `partname` together make up the `filename` in the "part 1" and "part 2" documents, default is `archivename`= `filename` and no `partname`
- `additionalArtifacts` contains references to other `oasis-tcs` repos, must occur only once per `archivename`

Fixes #254

### CODE REVIEW

File: odata-protocol/meta.yaml [hunk 1]
```diff
@@ -6,6 +6,13 @@ previousStage: csd01
 pubdate: 28 February 2024
 pubdateISO: '2024-02-28'
 copyright: © OASIS Open 2024
+archivename: odata-v4.02-csd02
+partname: part1-protocol
```
Comment (ralfhandl) [inline odata-protocol/meta.yaml:10]: Should we remove `filename` and use these two instead?

File: odata-url-conventions/meta.yaml [hunk 2]
```diff
@@ -6,6 +6,8 @@ previousStage: csd01
 pubdate: 28 February 2024
 pubdateISO: '2024-02-28'
 copyright: © OASIS Open 2024
+archivename: odata-v4.02-csd02
+partname: part2-url-conventions
```
Comment (ralfhandl) [inline odata-url-conventions/meta.yaml:10]: Should we remove `filename` and use these two instead?

File: odata-protocol/meta.yaml (hunk 1 above)
Comment (HeikoTheissen) [reply odata-protocol/meta.yaml:10]: No, a separate filename leaves more naming flexibility.

File: odata-url-conventions/meta.yaml (hunk 2 above)
Comment (ralfhandl) [reply odata-url-conventions/meta.yaml:10]: And needs more care in maintaining meta.yaml

File: lib/zip.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  if (meta.pubdateISO === pubDate) {
+    var archivename = meta.archivename || meta.filename;
+    archives[archivename] =
+      archives[archivename] || archiver("zip", { zlib: { level: 9 } });
+    archives[archivename].pipe(
+      fs.createWriteStream(
+        `${__dirname}/../zip/${archivename}-${meta.pubdateISO}.zip`,
+      ),
+    );
+    archives[archivename].directory(
```
Comment (ralfhandl) [inline lib/zip.js:18]: Rename md/html/pdf to use `meta.filename`

**Review by ralfhandl — CHANGES_REQUESTED** [review]

File: lib/zip.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+      for (var additional of meta.additionalArtifacts)
+        archives[archivename].append(
+          fs.createReadStream(
+            `${__dirname}/../../${additional.repo}/${additional.path}`,
+          ),
+          { name: additional.path },
+        );
+  }
+});
+for (var archivename in archives) archives[archivename].finalize();
```
Comment (ralfhandl) [inline lib/zip.js:32]:
```suggestion

for (var archivename in archives) {
  console.log(archivename);
  archives[archivename].finalize();
}
```

File: odata-url-conventions/meta.yaml (hunk 2 above)
Comment (HeikoTheissen) [reply odata-url-conventions/meta.yaml:10]: We use `filename` already in our markdown sources. OData-Protocol currently has
```yaml
filename: odata-v4.02-csd02-part1-protocol
archivename: odata-v4.02-csd02
partname: part1-protocol
```
What you propose is
```js
filename = archivename + (partname ?
  "-" + partname : "")
```
But this dynamic computation would have to be done by every consumer of the meta.yaml file.

Comment (HeikoTheissen) [reply odata-url-conventions/meta.yaml:10]: We could split the filename at the hyphen that is followed by "part" and avoid `archivename` and `partname`.

**Review by ralfhandl — APPROVED** [review]

File: lib/README.md [hunk 5]
```diff
@@ -96,6 +96,7 @@ The following scripts can be executed manually or as part of a GitHub Action:
   - a copy of the common [`styles`](../styles) folder
   - a copy of the document-specific `*/images` folder, if this exists.
 - [`npm run pdf`](build-pdf.mjs) runs the PDF conversion and writes the PDF document into the [`docs/*`](../docs) folder.
+- [`npm run zip <yyyy-MM-dd>`](zip.js) creates ZIP files for OASIS public review of documents with the given publication date. The ZIP files are written into the [`zip`](../zip) folder.
```
Comment (ralfhandl) [inline lib/README.md:99]: Why upper-case "MM"?

**Review by ralfhandl — DISMISSED** [review]

File: lib/README.md (hunk 5 above)
Comment (HeikoTheissen) [reply lib/README.md:99]: MM stands for month, mm stands for minutes.

Comment (ralfhandl) [reply lib/README.md:99]: Well, then why not YYYY-MM-DD, following https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date#date_time_string_format?

Which tool or standard uses yyyy-MM-dd?

Comment (HeikoTheissen) [reply lib/README.md:99]: > Which tool or standard uses yyyy-MM-dd?

https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html

**Comment by ralfhandl** [comment]
Feel free to merge

**Comment by ralfhandl** [comment]
One more thing: should we switch to the "release candidate" branch in the additional artifacts repos? And add this branch name to the `additionalArtifacts` objects?

**Comment by HeikoTheissen** [comment]
> One more thing: should we switch to the "release candidate" branch in the additional artifacts repos? And add this branch name to the `additionalArtifacts` objects?

I think no, because the current implementation simply copies the checked-out files, it does not perform any git commands (such as switching a branch).

**Review by ralfhandl — APPROVED** [review]

**Pull request merged**